hw/opentitan/ot_otp_eg.c

```diff
@@ -1564,58 +1610,109 @@ ot_otp_eg_load_partition_digest(OtOTPEgState *s, unsigned partition)
+
+    const uint8_t *base = (const uint8_t *)s->otp->data;
+    base += offset;
+
+    /* source address should be aligned to 64-bit boundary */
+    g_assert(((uintptr_t)base & (sizeof(uint64_t) - 1u)) == 0u);
+    const uint64_t *scrambled = (const uint64_t *)base;
+
+    /* destination address should be aligned to 64-bit boundary */
+    g_assert(pctrl->buffer.data != NULL);
+    uint64_t *clear = (uint64_t *)pctrl->buffer.data;
+
+    const uint8_t *scrambling_key = s->otp_scramble_keys[ix];
+    g_assert(scrambling_key);
+
+    OtPresentState *ps = ot_present_new();
+    ot_present_init(ps, scrambling_key);
+
+    trace_ot_otp_unscramble_partition(s->ot_id, PART_NAME(ix), ix, part_size);
+    /* neither the digest block nor the zeroizable block are scrambled */
+    for (unsigned dix = 0u; dix < dword_count; dix++) {
+        ot_present_decrypt(ps, scrambled[dix], &clear[dix]);
+    }
+
+    ot_present_free(ps);
+}
+
 static void ot_otp_eg_bufferize_partition(OtOTPEgState *s, unsigned ix)
 {
     OtOTPPartController *pctrl = &s->partctrls[ix];
 
     g_assert(pctrl->buffer.data != NULL);
 
     if (OtOTPPartDescs[ix].hw_digest) {
-        pctrl->buffer.digest = ot_otp_eg_load_partition_digest(s, ix);
+        pctrl->digest = ot_otp_eg_load_partition_digest(s, ix);
     } else {
-        pctrl->buffer.digest = 0;
+        pctrl->digest = 0;
     }
 
-    unsigned offset = (unsigned)OtOTPPartDescs[ix].offset;
-    unsigned part_size = OT_OTP_PART_DATA_BYTE_SIZE(ix);
+    if (OtOTPPartDescs[ix].secret) {
+        /* secret partitions need to be unscrambled */
+        if (s->blk) {
+            /*
+             * nothing to unscramble if no OTP data is loaded
+             * scrambling keys in this case may not be known
+             */
+            ot_otp_eg_unscramble_partition(s, ix);
+        }
+    } else {
+        unsigned offset = (unsigned)OtOTPPartDescs[ix].offset;
+        unsigned part_size = ot_otp_eg_part_data_byte_size(ix);
 
-    const uint8_t *base = (const uint8_t *)s->otp->data;
-    base += offset;
+        const uint8_t *base = (const uint8_t *)s->otp->data;
+        base += offset;
 
-    memcpy(pctrl->buffer.data, base, part_size);
+        memcpy(pctrl->buffer.data, base, part_size);
+    }
 }
 
-static void ot_otp_eg_check_partition_integrity(OtOTPEgState *s, unsigned ix)
+static void
+ot_otp_eg_check_buffered_partition_integrity(OtOTPEgState *s, unsigned ix)
 {
     OtOTPPartController *pctrl = &s->partctrls[ix];
 
-    if (pctrl->buffer.digest == 0) {
+    if (pctrl->digest == 0) {
         trace_ot_otp_skip_digest(s->ot_id, PART_NAME(ix), ix);
         pctrl->locked = false;
         return;
     }
 
     pctrl->locked = true;
 
-    unsigned part_size = OT_OTP_PART_DATA_BYTE_SIZE(ix);
+    /* buffered data has been eithed copied or unscrambled to the part buffer */
+    const uint8_t *part_data = (const uint8_t *)pctrl->buffer.data;
+    g_assert(part_data != NULL);
```

Comment:
This will cause failures for digest computation of secret partitions. The digests of secret partitions are calculated over the _scrambled_ data, rather than the _unscrambled_ data in the buffer.

The following fix was sufficient to get this behaviour working for me:

```suggestion
    /* digests are always calculated over the original data (scrambled or not) */
    const uint8_t *part_data =
        ((const uint8_t *)s->otp->data) + ot_otp_eg_part_data_offset(ix);
```

This is not necessarily the best fix - it assumes the OTP has not changed, and I have not checked whether in this instance HW will rescramble the scrambled buffer data instead (I suspect it might, but not sure if this is different between the *initialization logic* path and the *integrity check* path).

(repeated for DJ)

Reply:
I had a look, it does seem like HW will unscramble and then immediately rescramble the data for the initial integrity check, as it uses the same logical path for initialization and integrity checking:
* Upon initialization, after descrambling the last (non-digest) block, we [enter the `IntegDigClrSt` state](https://github.com/lowRISC/opentitan/blob/master/hw/ip_templates/otp_ctrl/rtl/otp_ctrl_part_buf.sv#L423).
* We also do this when idle [on any integrity check request](https://github.com/lowRISC/opentitan/blob/e383c23ab5d44940a07652ccd31406166286073d/hw/ip_templates/otp_ctrl/rtl/otp_ctrl_part_buf.sv#L480).
* In this integrity check (either initial or periodic), we must first [re-scramble](https://github.com/lowRISC/opentitan/blob/e383c23ab5d44940a07652ccd31406166286073d/hw/ip_templates/otp_ctrl/rtl/otp_ctrl_part_buf.sv#L577-L582) secret [partition data](https://github.com/lowRISC/opentitan/blob/e383c23ab5d44940a07652ccd31406166286073d/hw/ip_templates/otp_ctrl/rtl/otp_ctrl_part_buf.sv#L605-L609).
* After scrambling and computing the digest over the (re-)scrambled data, we check that the computed digest [matches the data in the partition's digest buffer registers](https://github.com/lowRISC/opentitan/blob/e383c23ab5d44940a07652ccd31406166286073d/hw/ip_templates/otp_ctrl/rtl/otp_ctrl_part_buf.sv#L691-L693).

So the patch above does not align the HW, but is correct because we only call `ot_otp_eg_check_buffered_partition_integrity` on initialization. If this function is re-used for the periodic integrity checks in the future, it will require this re-scrambling logic to be implemented.

Reply:
Ok thanks for double checking this. I do not thinkg periodic checking is something that is needed in QEMU, although it might be wanted at some (advanced) point to simulate failures and check how SW responds to them. That would be nice ... but definitely a far goal :-)
I think your fix is more that enough for the time being, thanks.